File: 300.500/488.ZumaGame.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <queue>
using namespace std;

// 记录列表的变化，剩余的颜色球，达到该变化消耗的步数
struct State
{
    string board{};
    string hand{};
    int step{};
    State(const string &b, const string &h, int s)
        : board(b), hand(h), step(s) {}
};

class ZumaGame
{
public:
    int findMinStep(string board, string hand)
    {
        // 记录已经处理过的情形:board和hand都相同，防止重复处理
        unordered_set<string> visited;
        // 对hand进行排序，将相同颜色的球排在一起，方便剪枝
        sort(hand.begin(), hand.end());

        // 开始广度优先遍历
        visited.insert(board + " " + hand);
        queue<State> qu;
        qu.push(State(board, hand, 0));
        while (!qu.empty())
        {
            State curr = qu.front();
            cout << "curr " << curr.board.c_str()
                 << " " << curr.hand.c_str() << " " << curr.step << endl;
            qu.pop();

            for (int j = 0; j < curr.hand.size(); ++j)
            {
                // 第1种剪枝情况: 当前球与上一个球颜色相同，不用重复处理
                if (j > 0 && curr.hand[j - 1] == curr.hand[j])
                {
                    continue;
                }
                for (int i = 0; i <= curr.board.size(); ++i)
                {
                    // 第2种剪枝情况: 在连续相同颜色的球，插入到那个位置是一样的，因此只需要处理插入在头的情况
                    if (i > 0 && curr.board[i - 1] == curr.hand[j])
                    {
                        continue;
                    }
                    // 第3种剪枝情况:只在这两种情况插入球
                    bool choose = false;
                    // 第3_1种情况：当前球颜色与后面的球的颜色相同
                    if (i < curr.board.size() && curr.board[i] == curr.hand[j])
                    {
                        choose = true;
                    }
                    // 第3_2种情况 : 当前后颜色相同且与当前颜色不同时候放置球
                    if (i > 0 && i < curr.board.size() && curr.board[i - 1] == curr.board[i] && curr.board[i] != curr.hand[j])
                    {
                        choose = true;
                    }
                    if (choose)
                    {
                        string newBoard = clean(curr.board.substr(0, i) + curr.hand[j] + curr.board.substr(i));
                        string newHand = curr.hand.substr(0, j) + curr.hand.substr(j + 1);
                        if (newBoard.size() == 0)
                        {
                            return curr.step + 1;
                        }
                        if (!visited.count(newBoard + " " + newHand))
                        {
                            qu.push(State(newBoard, newHand, curr.step + 1));
                            visited.insert(newBoard + " " + newHand);
                        }
                    }
                }
            }
        }

        return -1;
    }

    int findMinStep1(string board, string hand)
    {
        sort(hand.begin(), hand.end());
        dp.clear();
        int ans = dfs(board, hand);
        return ans <= 5 ? ans : -1;
    }

private:
    /* 用于清除string中达到或超过3个的连续相同颜色
    先建立<颜色，连续个数>的列表
    再将 “连续个数” >= 3 的全部移除
    最后重新构建字符串
    */
    string clean(const string &s)
    {
        vector<pair<char, int>> st;
        for (char c : s)
        {
            // 移除已经满3个的 （必须与当前char不同，否则应该累加数量）
            while (!st.empty() && c != st.back().first && st.back().second >= 3)
            {
                st.pop_back();
            }
            if (st.empty() || c != st.back().first)
            {
                // 加入新元素
                st.push_back({c, 1});
            }
            else
            {
                st.back().second++;
            }
        }
        // 最有再检测一遍
        if (!st.empty() && st.back().second >= 3)
        {
            st.pop_back();
        }
        // 原有字符串
        string res = "";
        for (int i = 0; i < st.size(); ++i)
        {
            for (int j = 0; j < st[i].second; ++j)
            {
                res.push_back(st[i].first);
            }
        }
        return res;
    }

    int dfs(const string &board, const string &hand)
    {
        if (board.size() == 0)
        {
            return 0;
        }
        // 已经检测过了，不再重复检测
        if (dp.count(board + " " + hand))
        {
            return dp[board + " " + hand];
        }
        int res = 6;
        for (int j = 0; j < hand.size(); ++j)
        {
            // 第1种剪枝情况: 当前球与上一个球颜色相同，不用重复处理
            if (j > 0 && hand[j] == hand[j - 1])
            {
                continue;
            }
            for (int i = 0; i <= board.size(); ++i)
            {
                // 第2种剪枝情况: 在连续相同颜色的球，插入到那个位置是一样的，因此只需要处理插入在头的情况
                if (i > 0 && board[i - 1] == hand[j])
                {
                    continue;
                }
                // 第3种剪枝情况:只在这两种情况插入球
                bool choose = false;
                // 第3_1种情况：当前球颜色与后面的球的颜色相同
                if (i < board.size() && board[i] == hand[j])
                {
                    choose = true;
                }
                // 第3_2种情况 : 当前后颜色相同且与当前颜色不同时候放置球
                if (i > 0 && i < board.size() && board[i - 1] == board[i] && board[i] != hand[j])
                {
                    choose = true;
                }
                if (choose)
                {
                    string newBoard = clean(board.substr(0, i) + hand[j] + board.substr(i));
                    string newHand = hand.substr(0, j) + hand.substr(j + 1);
                    res = min(res, dfs(newBoard, newHand) + 1);
                }
            }
        }
        dp[board + " " + hand] = res;
        return res;
    }

private:
    unordered_map<string, int> dp;
};

int main(int argc, char const *argv[])
{
    ZumaGame cls;
    cout << cls.findMinStep1("RRWWRRBBRR", "WB") << " -> 2" << endl;
    cout << cls.findMinStep("BGGRRYY", "BBYRG") << " -> 5" << endl;
    cout << cls.findMinStep1("WRRBBW", "RB") << " -> -1" << endl;
    cout << cls.findMinStep1("WWRRBBWW", "WRBRW") << " -> 2" << endl;
    cout << cls.findMinStep1("G", "GGGGG") << " -> 2" << endl;
    cout << cls.findMinStep1("RBYYBBRRB", "YRBGB") << " -> 3" << endl;
    return 0;
}

File: go/300.500/488.ZumaGame.go
package main

import (
	"bytes"
	"fmt"
	"sort"
)

func min(i, j int) int {
	if i < j {
		return i
	}
	return j
}

type State struct {
	board []byte
	hand  []byte
	step  int
}

func NewState(b, h []byte, s int) *State {
	return &State{
		board: b,
		hand:  h,
		step:  s,
	}
}

func findMinStep(bs string, hs string) int {
	visited := map[string]struct{}{}
	board := []byte(bs)
	hand := []byte(hs)
	sort.Slice(hand, func(i, j int) bool {
		return hand[i] <= hand[j]
	})

	visited[bs+" "+hs] = struct{}{}
	queue := []*State{}
	queue = append(queue, NewState(board, hand, 0))
	for len(queue) > 0 {
		curr := queue[0]
		// fmt.Println("curr", string(curr.board), string(curr.hand), curr.step)
		queue = queue[1:]
		for j := 0; j < len(curr.hand); j++ {
			if j > 0 && curr.hand[j-1] == curr.hand[j] {
				continue
			}
			for i := 0; i <= len(curr.board); i++ {
				if i > 0 && curr.hand[j] == curr.board[i-1] {
					continue
				}
				choose := false
				if i < len(curr.board) && curr.board[i] == curr.hand[j] {
					choose = true
				}
				if i > 0 && i < len(curr.board) && curr.board[i-1] == curr.board[i] && curr.board[i] != curr.hand[j] {
					choose = true
				}
				if choose {
					newSource := bytes.Join([][]byte{curr.board[:i], []byte{curr.hand[j]}, curr.board[i:]}, []byte{})
					newBoard := clean(newSource)
					if len(newBoard) == 0 {
						return curr.step + 1
					}
					newHand := bytes.Join([][]byte{curr.hand[:j], curr.hand[j+1:]}, []byte{})
					key := string(newBoard) + " " + string(newHand)
					// fmt.Println("check", key)
					_, ok := visited[key]
					if !ok {
						queue = append(queue, NewState(newBoard, newHand, curr.step+1))
						visited[key] = struct{}{}
					}
				}
			}
		}
	}
	return -1
}

type pair struct {
	char  byte
	count int
}

func clean(s []byte) []byte {
	//fmt.Println("in", string(s))
	st := []*pair{}
	for _, c := range s {
		for {
			size := len(st)
			if size == 0 {
				break
			}
			pa := st[size-1]
			if c == pa.char {
				break
			}
			if pa.count < 3 {
				break
			}
			st = st[:size-1]
		}
		if len(st) == 0 || c != st[len(st)-1].char {
			st = append(st, &pair{
				char:  c,
				count: 1,
			})
		} else {
			st[len(st)-1].count += 1
		}
	}
	if len(st) > 0 && st[len(st)-1].count >= 3 {
		st = st[:len(st)-1]
	}
	res := []byte{}
	for _, v := range st {
		for j := 0; j < v.count; j++ {
			res = append(res, v.char)
		}
	}
	//fmt.Println("out", string(res))
	return res
}

var dp map[string]int

func findMinStep1(bs string, hs string) int {
	dp = map[string]int{}
	board := []byte(bs)
	hand := []byte(hs)
	sort.Slice(hand, func(i, j int) bool {
		return hand[i] <= hand[j]
	})
	ans := dfs(board, hand)
	if ans <= 5 {
		return ans
	}
	return -1
}

func dfs(board, hand []byte) int {
	if len(board) == 0 {
		return 0
	}
	key := fmt.Sprintf("%s %s", board, hand)
	if v, ok := dp[key]; ok {
		return v
	}
	res := 6
	for j := 0; j < len(hand); j++ {
		if j > 0 && hand[j] == hand[j-1] {
			continue
		}
		for i := 0; i < len(board); i++ {
			if i > 0 && board[i-1] == hand[j] {
				continue
			}
			choose := false
			if i < len(board) && board[i] == hand[j] {
				choose = true
			} else if i > 0 && i < len(board) && board[i-1] == board[i] && board[i] != hand[j] {
				choose = true
			}
			if choose {
				newSource := bytes.Join([][]byte{board[:i], []byte{hand[j]}, board[i:]}, []byte{})
				newBoard := clean(newSource)
				newHand := bytes.Join([][]byte{hand[:j], hand[j+1:]}, []byte{})
				res = min(res, dfs(newBoard, newHand)+1)
			}
		}
	}
	dp[key] = res
	return res
}

func main() {
	fmt.Println(findMinStep1("RRWWRRBBRR", "WB"), 2)
	fmt.Println(findMinStep1("RRGGBBYYWWRRGGBB", "RGBYW"), -1)
	fmt.Println(findMinStep1("BGGRRYY", "BBYRG"), 5)
	fmt.Println(findMinStep1("RBYYBBRRB", "YRBGB"), 3)
	fmt.Println(findMinStep1("WRRBBW", "RB"), -1)
	fmt.Println(findMinStep1("WWRRBBWW", "WRBRW"), 2)
	fmt.Println(findMinStep1("G", "GGGGG"), 2)
}
